File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryReaderUtil.java
package org.cqframework.cql.cql2elm;

import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.Reader;

import javax.xml.transform.Source;
import javax.xml.transform.stream.StreamSource;


public class LibraryReaderUtil {

    /**
     * Creates {@link Source} from various JSON representation.
     */
    public static Source toSource(Object json) throws IOException {
        if (json == null)
            throw new CqlTranslatorException("no JSON is given");

        if (json instanceof String) {
            try {
                json = new URI((String)json);
            } catch (URISyntaxException e) {
                json = new File((String)json);
            }
        }

        if (json instanceof File) {
            return new StreamSource((File)json);
        }

        if (json instanceof URI) {
            json = ((URI)json).toURL();
        }

        if (json instanceof URL) {
            return new StreamSource(((URL)json).toExternalForm());
        }

        if (json instanceof InputStream) {
            return new StreamSource((InputStream)json);
        }

        if (json instanceof Reader) {
            return new StreamSource((Reader)json);
        }

        if (json instanceof Source) {
            return (Source)json;
        }

        throw new CqlTranslatorException(String.format("Could not determine access path for input of type %s.", json.getClass()));
    }
}

File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryReaderUtil.kt
@file:Suppress("WildcardImport")

package org.cqframework.cql.cql2elm

import java.io.*
import java.net.URI
import java.net.URISyntaxException
import java.net.URL
import javax.xml.transform.Source
import javax.xml.transform.stream.StreamSource

object LibraryReaderUtil {
    /** Creates [Source] from various JSON representation. */
    @Suppress("ReturnCount")
    @Throws(IOException::class)
    fun toSource(json: Any?): Source {
        var json: Any? = json ?: throw CqlCompilerException("no JSON is given")
        if (json is String) {
            json =
                try {
                    URI(json)
                } catch (@Suppress("SwallowedException") e: URISyntaxException) {
                    File(json)
                }
        }
        if (json is File) {
            return StreamSource(json as File?)
        }
        if (json is URI) {
            json = json.toURL()
        }
        if (json is URL) {
            return StreamSource(json.toExternalForm())
        }
        if (json is InputStream) {
            return StreamSource(json as InputStream?)
        }
        if (json is Reader) {
            return StreamSource(json as Reader?)
        }
        if (json is Source) {
            return json
        }
        throw CqlCompilerException(
            @Suppress("ImplicitDefaultLocale")
            String.format("Could not determine access path for input of type %s.", json!!.javaClass)
        )
    }
}
